Passive scalar

gloopydoop opened the pull request:
don't review this yet I'm checking the linting!!

timfelle commented on sources/objectives/minimum_dissipation_objective_function.f90:
```diff
@@ -255,22 +262,25 @@ end subroutine minimum_dissipation_objective_function_compute
 
   !> compute the sensitivity of the objective function with respect to $\chi$
```
```suggestion
  !> compute the sensitivity of the objective function with respect to \f$\chi\f$
```
Doxygen does not directly understand `$` to mean a function. You have to tell it by using `\f$` instead.

timfelle committed:
Merge branch 'develop' into example/brinkman_study

timfelle committed:
Update to new bc list interface

timfelle committed:
Improve handling install and working folders.

additionally stop reporting warnings from external dependencies

timfelle committed:
Prevent deletion of work folder during cleanup (#70)

Ensure the work folder is not deleted if the install directory is
contained within it during the cleanup process.

timfelle committed:
Merge commit '48b0b8ab4e5f09024c8b04922fe0685e018ea099' into fix/bc_interface

timfelle committed:
Update to new bc list interface (#69)

timfelle committed:
Merge remote-tracking branch 'origin/develop' into enhancement/design

timfelle committed:
Objective now initialize with a simulation. and compute with just a design

timfelle committed:
cleanup handling of masks

timfelle committed:
added a factory for objective functions

timfelle committed:
extensive cleanup

timfelle committed:
Using the generic design type

timfelle committed:
Sync with the working branch

timfelle committed:
Update case

timfelle committed:
revert breaking change

timfelle committed:
remove merge mistake from cmake lists

timfelle committed:
Slight cleanup

timfelle committed:
Introduce objectives and constraints

timfelle committed:
Objectives and constraints are components of problems

timfelle committed:
fix bug in cmake lists

timfelle committed:
Slight cleanup

AbbasMoosaV committed:
MMA parameters added to json

AbbasMoosaV committed:
The optimizer section of README.md is updated

AbbasMoosaV committed:
Merge branch 'develop' into feature/optimizer

AbbasMoosaV committed:
Addressed PR comments: The old rugby ball example is removed. The max_iter number for the optimization loop is now read from json

AbbasMoosaV committed:
Addressed PR comments: optimizer and mma are moved to optimizer folder, Cmakelist.txt is updated, undo the changes to steady_state_problem.f90

AbbasMoosaV committed:
Addressed PR comments: the pointer to problem_t is removed from the optimizer; the problem is now given as an input the the functions init and run in the optimizer module.

AbbasMoosaV committed:
Update sources/CMakeLists.txt

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Update examples/rugby_ball/driver.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Update examples/rugby_ball/driver.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Update sources/optimizer/mma.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Update examples/rugby_ball/driver.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Update sources/optimizer/optimizer.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Update sources/optimizer/mma_optimizer.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Update sources/optimizer/mma_optimizer.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Update sources/optimizer/mma.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Update sources/optimizer/mma_optimizer.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Update sources/optimizer/mma_optimizer.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Update sources/optimizer/mma_optimizer.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Update sources/optimizer/mma_optimizer.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Addressed PR comments: scaling moved to mma and some minor comments implemented

AbbasMoosaV committed:
Merge branch 'feature/optimizer' of github.com:ExtremeFLOW/neko-top into feature/optimizer

AbbasMoosaV committed:
Fix some linting issues.

AbbasMoosaV committed:
Addressed PR comments: update the optimization_data output & a few cleanups.

timfelle committed:
Feature/optimizer (#71)

- An optimizer type is added to improve the structure of the topology
optimization solver.
- MMA.f90 is updated to read the required parameters from the json case
file.
- Two different ways of scaling the objective and its derivatives is
considered.
- The rugby_ball example is updated to rugby_ball_optimizerobj to be use
the optimizer type.
- The old rugby_ball example is no longer supported due to changes in
MMA.f90.
to do list:
- The ...%free() functions needs to be checked and updated.

timfelle committed:
started work on cleaning the design

timfelle committed:
Add vector version of masked exterior

timfelle committed:
Split the simulation

timfelle committed:
Add size function

timfelle committed:
Use the vector exterior masking

timfelle committed:
prepare interface

timfelle committed:
next step on moving the interface

timfelle committed:
Still works when calling the functions

timfelle committed:
use getters and stop memory leak

timfelle committed:
set number of constraints and such in problem

timfelle committed:
add the sensitivities interface

timfelle committed:
Hide the objectives and constraints

timfelle committed:
unhide the changes

timfelle committed:
Merge commit 'f664ac4356d8cdc23a5c1cf5839b131d1929c8bb' into cleanup

timfelle committed:
Started cleanup of mma interface

timfelle committed:
MMA interface aligned with new problem

timfelle committed:
Remove redundant prints

timfelle committed:
Add objectives and constraints as a list

timfelle committed:
Add a weight to the objectives

timfelle committed:
Update base initializer.

timfelle committed:
small updates

timfelle committed:
Rename toggle for masks and updated initializers

timfelle committed:
Introduce a lube term. Not clean yet.

timfelle committed:
Cleanup of the minimal dissipation

timfelle committed:
Cleanup of lube term objective

timfelle committed:
update intent on json

timfelle committed:
Json can now be read

timfelle committed:
bit more cleanup

timfelle committed:
Cleanup the volume

timfelle committed:
Update documentation

timfelle committed:
Cleanup the volume some more

timfelle committed:
slight cleanup

timfelle committed:
Add n constraints as inputs for MMA

timfelle committed:
Updated documentation of problem

timfelle committed:
Slightly more cleanup

timfelle committed:
Merge commit 'f664ac4356d8cdc23a5c1cf5839b131d1929c8bb' into example/brinkman_study

timfelle committed:
Enhance Brinkman flow examples and metrics computation (#64)

This update adds the examples for Brinkman flow, adds new metrics
computation features, and enhances the overall structure and robustness
of the code. It includes updates to the generator script, improved error
handling, and additional documentation for better usability.

Depends on: 

- [x] #67 

Closes #60

timfelle committed:
Merge branch 'develop' into enhancement/design

timfelle committed:
Handling of the problem type (#52)

This PR exposes some of the work we are doing in improving the problem
type.
The hope is to make it relatively generic and minimize which thing need
to be specified by a derived type.

This PR will primarily touch:

- How an optimization problem is defined (problem.f90, and derived
types)
- How the problem interact with a "design" or solution. (design.f90, and
detived types)
- Representation and handling of constraints and objectives.

timfelle committed:
Add verification example

timfelle committed:
Revert "Handling of the problem type (#52)"

This reverts commit dd9eb66cf2dc2df7e481df9fd9d72e72a58e469f, reversing
changes made to 841240ac439cdb742d058637dfd5fd18b9b41863.

gloopydoop committed:
updated the jupyter notebook to what we had in the deliverable

timfelle committed:
Add verification example and revert previous handling changes (#73)

Introduce a verification example while reverting changes related to
problem type handling to restore previous functionality.

gloopydoop committed:
merge develop and update passive scalar BCs

timfelle committed:
Fix the design alone

gloopydoop committed:
new BCs

timfelle committed:
Add the simulation type

timfelle committed:
A few misc operator updates

gloopydoop committed:
midway through merging develop

timfelle committed:
update mask operator

timfelle committed:
Update the steady sim-comp

timfelle committed:
Update adjoint source term for lube

timfelle committed:
limit to single number of procs to speed up testing

timfelle committed:
Move files to new locations

timfelle committed:
move the objective file

timfelle committed:
Add base class for functionals

timfelle committed:
Introduce the constraint and factory

timfelle committed:
Fix precision bug which invalidates all results

gloopydoop committed:
omg it's still fucked

timfelle committed:
Targets are machine dependent

timfelle committed:
add a simulation instead of cases

timfelle committed:
Formatting

timfelle committed:
Start to align interfaces

timfelle committed:
rename problem

timfelle committed:
update optimizer interface

timfelle committed:
Remove design pointer from problem

timfelle committed:
eliminate direct pointers to the cases

timfelle committed:
Set loop limits for the subsolve

timfelle committed:
Grab the rest of the updates

timfelle committed:
add bc factory, just a copy of fluid

timfelle committed:
fix of pde filter bcs

timfelle committed:
Started to adjust for the overhaul

timfelle committed:
very broken still

timfelle committed:
Add a few markers indicating what should be moved

timfelle committed:
Add more markers of things to move

timfelle committed:
interface alignment and adding files to cmake

timfelle committed:
Updates to the code to align schemes

timfelle committed:
progress

timfelle committed:
Add missing definition

timfelle committed:
Add missing CUDA definition (#78)

Include a compile definition for CUDA support when the DEVICE_TYPE is
set to CUDA.

timfelle committed:
MPI fixed by encapsulating in_place variable in subroutines

timfelle committed:
We compile

timfelle committed:
revert advection updates

timfelle committed:
compiles

timfelle committed:
Forward seem alright again

timfelle committed:
adjoint bc update

timfelle committed:
So far so good

timfelle committed:
Slightly more cleanup

timfelle committed:
updated the jupyter notebook to what we had in the deliverable (#74)

I didn't realize this files was only on my computer locally.

timfelle committed:
Apply suggestions from code review

Co-authored-by: Harrison Nobis <58177854+gloopydoop@users.noreply.github.com>

timfelle committed:
Apply suggestion from abbas

timfelle committed:
Update to sort of fix the example

timfelle committed:
Update the github actions

timfelle committed:
update the formatting workflow

timfelle committed:
Do not warn about missing include directories

timfelle committed:
Update the main workflow.

- We now cache all external dependencies as separate jobs.
- Formatting is enforced on modified files.

timfelle committed:
Add missing dependencies

timfelle committed:
Directly assign the flags instead of relying on a variable

timfelle committed:
Update the formatter

timfelle committed:
Fix/bc_overhaul (#77)

timfelle committed:
Merge branch 'develop' into ci/merge_group

timfelle committed:
Some caching did not work

timfelle committed:
Merge commit '44e77d1448e1580815d0240032c5bd63d28caf2e' into fix/design

timfelle committed:
update verification example driver

timfelle committed:
Test working on a merge queue (#80)

A few updates to the CI. Some of this is to support the merge queues by
github.
The merge queue ensure multiple PR's being accepted at roughly the same
time does not break the repository. Each accepted Pr is added to a queue
and merged in order. This should help streamline our development, while
every merge is tested before the next is added.

- [x] Develop should properly support the merge queues.
- [x] The formatting action is now functional.
- [x] Compilation check should fail on warnings, such as un-initialized
variables.

timfelle committed:
Merge remote-tracking branch 'origin/develop' into fix/design

timfelle committed:
Minor update

timfelle committed:
cleanup flow diagram for workflow

timfelle committed:
Re introduction of the design class (#76)

The bugs have now ben fixed and this is a reintroduction of the updated
design types

Depends on #77

timfelle committed:
Synchronize the formatting action with Neko

timfelle committed:
Add some details on the formatter

timfelle committed:
update example

timfelle committed:
Add missing precisions

timfelle committed:
fix a bug causing a MPI de-synchronization

timfelle committed:
Cleanup round 1

timfelle committed:
Cleanup round 2, solution updated

timfelle committed:
Cleanup 3,

timfelle committed:
Cleanup 4

timfelle committed:
remove some unused settings

timfelle committed:
Cleanup 5

timfelle committed:
Cleanup 6

timfelle committed:
Hack support for the new solver updates

timfelle committed:
fix format

timfelle committed:
Hack support for the new solver updates (#83)

Enhance compatibility with the latest solver updates by modifying fluid
scheme references and ensuring proper initialization across various
subroutines.

timfelle committed:
Merge branch 'develop' into fix/mma_cpu

timfelle committed:
Merge branch 'develop' into ci/format

timfelle committed:
apply review comments

timfelle committed:
Synchronize formatting action with Neko (#81)

Align the formatting process with Neko's standards and update the
contribution guidelines to reference Neko's documentation.

timfelle committed:
Merge remote-tracking branch 'origin/develop' into fix/mma_cpu

timfelle committed:
Fix MPI synchronization bug and clean up codebase (#82)

Address a bug that caused MPI de-synchronization and enhance code
clarity through multiple cleanup rounds. Additionally, update example
configurations and refine precision settings.

timfelle committed:
Fix buggy imports

timfelle committed:
Fix buggy imports (#84)

Refactor module imports to ensure proper usage and avoid conflicts. This
improves code clarity and maintains consistency across modules.

timfelle committed:
Discussions with Niclas led to some more finds in terms of issues.

timfelle committed:
Work on array temporaries

timfelle committed:
Fix preprocessor work.
Fix cmake bug in how compile definition was applied.

timfelle committed:
Ignore unneeded flags again

timfelle committed:
Fix CMake compile definition and preprocessor directives for CUDA (#87)

Correct the application of the compile definition for CUDA in CMake and
update preprocessor directives to ensure proper conditional compilation.

timfelle committed:
Enhance Fortran build configuration and add LAPACK interface (#86)

Improve debugging options in the Fortran build configuration and
introduce a new LAPACK interface module for solving linear equations.
Additionally, refine type declarations for better clarity in the code.

gloopydoop committed:
move mapping

gloopydoop committed:
stamping

gloopydoop committed:
how did I fail the linter on osmething so small...

timfelle committed:
Relax the debug

timfelle committed:
Revert "Enhance Fortran build configuration and add LAPACK interface (#86)"

This reverts commit 1448f00ad375d8172e08f8ce66cd847b2a92bd8f, reversing
changes made to 4c836a86deccc3f3f1e4609a1492fd9c0cc8ac2c.

gloopydoop committed:
merged, really poorly

timfelle committed:
Discussions with Niclas led to some more finds in terms of issues.

(cherry picked from commit f56db72a933cd10ef764e80395a6d4b30af3b055)

timfelle committed:
Work on array temporaries

timfelle committed:
Force run the CI for debug purpose

timfelle committed:
Don't cache system for pfunit

timfelle committed:
same for gslib

timfelle committed:
Reactivate cache and improve parallelism

timfelle committed:
Move checks about a bit, but it should be good now

timfelle committed:
Commenting update to improve readability of workflow

timfelle committed:
forgot to remove check from main workflow

timfelle committed:
test without cache on neko

timfelle committed:
Once again

timfelle committed:
Reactivate the cache

timfelle committed:
un-cache json

timfelle committed:
skip caching for system

timfelle committed:
reactivate cache

timfelle committed:
Apt cache is broken

timfelle committed:
Revert Fortran build configuration changes and improve debugging options (#89)

Revert previous enhancements to the Fortran build configuration while
relaxing debug options and addressing identified issues with array
temporaries.

timfelle committed:
MMA loop tweeks (#88)

Very small PR, but pretty important.

The design should be mapped to the brinkman term in between the MMA
update and running the simulation.

Also, stamping the zeroth iteration is nice for visualization.

timfelle committed:
Massive improvement to the cmake structure.

- Neko-TOP static library is built by default
- The `Examples` target now builds the examples.
- The setup script is updated to account for this.

timfelle committed:
Improve CMake structure and build process for examples (#90)

- Neko-TOP static library is built by default
- The `Examples` target now builds the examples.
- The setup script is updated to account for this.

Time to run `./setup.sh` (With all external dependencies ready)

- Before: 41 sec
- After: 26 sec

gloopydoop committed:
first draft

gloopydoop committed:
little adjustment

timfelle committed:
Declare in same order as usage in the files

timfelle committed:
Move MMA settings into the solver section

timfelle committed:
Small clean up of optimizer

timfelle committed:
Check against the result folder instead of a log

timfelle committed:
Add a factory for generic optimizer

timfelle committed:
Fix uninitialized variables

timfelle committed:
Update driver with the parameter read

timfelle committed:
Simplify optimizer (Invalidated previous results)

timfelle committed:
Updates are functional

timfelle committed:
Slight cleanup

timfelle committed:
Add base initializer

timfelle committed:
Updates to our interfaces

timfelle committed:
Align the interfaces with style from neko

timfelle committed:
Improve the driver structure.

timfelle committed:
Use the simulation reset function

timfelle committed:
Update parameter order

timfelle committed:
Add a file reader function for json

gloopydoop committed:
fromatting

gloopydoop committed:
slight updates

gloopydoop committed:
minor changes

gloopydoop committed:
csv_file_t

gloopydoop committed:
PR comments

gloopydoop committed:
lint

gloopydoop committed:
PR comments

gloopydoop committed:
.txt -> .csv

gloopydoop committed:
formatting

gloopydoop committed:
back to the zeroth iteration

gloopydoop committed:
initial condition

timfelle committed:
Improved logging (#91)

I guess we discussed this a bit this morning, but this is what I had in
mind.
I'm putting this in draft mode because I want comments before I move any
further with this:

- Do we think we'll want this in other optimization? ie, would a
`logger_t` be useful?
- Are we all happy with enforcing a log name to all
constraints/objectives so we can write them out?

But perhaps more importantly, what I've written currently is pretty
susceptible to truncation of that long string (buffer? is this the
correct word? that was the intention of the variable name `buff`)

### Example output

```
n =     86400, m =         1, tolerance =  1.00000000000000002E-03
iter, Total objective function, Dissipation, out of plane stresses, Volume constraint, KKTmax, KKTnorm2, scalaing factor
  0,     5.3696209634E+01,     4.0441625841E+01,     1.3254583793E+01,    -3.4113716238E-01,     1.7976931349+308,     1.7976931349+308,     1.0000000000E+00
  1,     5.3631210218E+01,     4.0441625841E+01,     1.3189584378E+01,    -3.4067827356E-01,     3.4063388244E-01,     1.8251180684E+00,     1.0000000000E+00
  2,     5.3266158113E+01,     4.0127814834E+01,     1.3138343279E+01,    -3.4021952588E-01,     3.4017508569E-01,     5.0791113981E+00,     1.0000000000E+00
  3,     5.2946050791E+01,     3.9894493099E+01,     1.3051557691E+01,    -3.3973026153E-01,     3.3968611728E-01,     4.1537413064E+00,     1.0000000000E+00
  4,     5.2488335978E+01,     3.9561169448E+01,     1.2927166530E+01,    -3.3922171690E-01,     3.3917795103E-01,     5.6123647507E+00,     1.0000000000E+00
  5,     5.1825567047E+01,     3.9082081599E+01,     1.2743485448E+01,    -3.3889046816E-01,     3.3884813432E-01,     7.9908112190E+00,     1.0000000000E+00
  6,     5.1000305086E+01,     3.8491289553E+01,     1.2509015532E+01,    -3.3852679847E-01,     3.3848677945E-01,     9.8497959414E+00,     1.0000000000E+00
  7,     5.0354702281E+01,     3.8025030555E+01,     1.2329671726E+01,    -3.3866646158E-01,     3.3862891930E-01,     7.8845237703E+00,     1.0000000000E+00
  8,     4.9785292383E+01,     3.7615237248E+01,     1.2170055135E+01,    -3.3902719413E-01,     3.3899308096E-01,     7.5623243319E+00,     1.0000000000E+00

```

### Example output using `csv_file_type`
```
iter, Total objective function, Dissipation, Out of plane stresses, Volume constraint, KKTmax, KKTnorm2, scalaing factor
0.0000000000000000,53.696209634065141,40.441625840817053,13.254583793248088,-0.34113716238378983,0.0000000000000000,0.0000000000000000,1.0000000000000000
1.0000000000000000,53.631210218451116,40.441625840817053,13.189584377634059,-0.34067827356064134,0.34063388243730330,1.8251180684119177,1.0000000000000000
2.0000000000000000,53.266158113095081,40.127814833617478,13.138343279477603,-0.34021952587850796,0.34017508568651300,5.0791113980976883,1.0000000000000000
3.0000000000000000,52.946050790610592,39.894493099386999,13.051557691223593,-0.33973026152689706,0.33968611727815884,4.1537413063900503,1.0000000000000000
4.0000000000000000,52.488335977948438,39.561169447724346,12.927166530224092,-0.33922171690237829,0.33917795102621462,5.6123647507375267,1.0000000000000000
5.0000000000000000,51.825567047220396,39.082081599088639,12.743485448131759,-0.33889046815908813,0.33884813432446081,7.9908112190203324,1.0000000000000000
6.0000000000000000,51.000305085635944,38.491289553207920,12.509015532428020,-0.33852679846645700,0.33848677945129146,9.8497959414055956,1.0000000000000000
7.0000000000000000,50.354702281281916,38.025030555367294,12.329671725914622,-0.33866646158438390,0.33862891930457578,7.8845237703490136,1.0000000000000000
8.0000000000000000,49.785292383215989,37.615237248149093,12.170055135066898,-0.33902719412641674,0.33899308096021136,7.5623243318792319,1.0000000000000000
```

timfelle committed:
Merge commit '3f11b506ff4b802167a10aafb58841c947ea526f' into driver/json

timfelle committed:
Merge commit '7a0b303bcd03c05e12edd54a976826a89824f2b2' into driver/json

timfelle committed:
Fix the possibility to specify the number of processors to run with

timfelle committed:
Allow specification of processor count for execution (#92)

Introduce the ability to specify the number of processors to use via the
`NPROCS` environment variable, enhancing flexibility in resource
allocation during script execution.

gloopydoop committed:
still working

timfelle committed:
Enhance JSON handling and driver structure improvements (#72)

Introduce json based initialization for all main types.

Additionally I have split the simulation out of the problem in order to
simplify dependencies.

Depends on #76

gloopydoop committed:
doing so poorly on the merge

gloopydoop committed:
worst merge ever, starting from scratch

gloopydoop committed:
still in a weird midway point

gloopydoop committed:
WIP
